Bluetooth: Mesh: Flush pending settings in bt_mesh_reset/_cdb_clear

PavelVPV opened the pull request:
Upon bt_mesh_reset call, all mesh modules has to clear the privisioning
and configuration data. None of the modules store or erase in settings
subsystem immediately, but put this on the mesh settings work. After the
settings work is scheduled, all stored data will be removed and the
device will eventually be unprovisioned. Until then, the device is not
completely unprovisioned, thus calling bt_mesh_prov_enable,
bt_mesh_provision (and bt_mesh_cdb_create) should not be allowed.

Struct bt_mesh_prov has a reset callback stating that after this
callback is called, the device has been reset and can be re-provisioned
again. Also, bt_mesh_reset API description states that after calling
bt_mesh_reset API, the device needs to reenable the provisioning layer
to be provisioned again. But this is not correct since the settings has
to be cleared before the device can be reprovisionined.

This commit makes bt_mesh_reset flush pending settings so that the
device can be reprovisioned immediately and to API will behave as
written in the description.

The same applies to bt_mesh_cdb_clear.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Andrewpini commented:
Nitpick: 
![image](https://user-images.githubusercontent.com/35562897/178421667-fd63c394-dbcb-4190-a394-0a62d44111af.png)
"the"?

PavelVPV commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -2071,13 +2078,13 @@ static void reset_send_start(uint16_t duration, int err, void *cb_data)
 	if (err) {
 		BT_ERR("Sending Node Reset Status failed (err %d)", err);
-		bt_mesh_reset();
+		k_work_submit(&node_reset_pending);
```
It runs on the advertiser thread, but requires more stack size than allocated for this thread.

alxelax replied:
Hm. How much larger is the memory required? Will it be simpler just to add stack?

IMHO Such kind of optimizations is a bit strange. This is attempt to decrease memory consumption by some OS primitives usage that is designed for thread scheduling. 
I do not strictly agree or disagree but it is a kind of hidden knowledge that is difficult to follow up.

PavelVPV replied:
With bt_mesh_reset() flushing pending settings in bt_mesh_send_cb.end:

```
uart:~$ mesh init
Mesh shell initialized
uart:~$ kernel stacks
0x20002460 BT Mesh adv (real size 2000):        unused 1312     usage 688 / 2000 (34 %)
...
The local node has been reset and needs reprovisioning
uart:~$ kernel stacks
0x20002460 BT Mesh adv (real size 2000):        unused 1160     usage 840 / 2000 (42 %)
```

Sizeof k_work is 16 bytes.
If you think it is wise to increase adv stack size, I can do this.

PavelVPV committed:
Bluetooth: Mesh: Flush pending settings in bt_mesh_reset/_cdb_clear

Upon bt_mesh_reset call, all mesh modules has to clear the privisioning
and configuration data. None of the modules store or erase in settings
subsystem immediately, but put this on the mesh settings work. After the
settings work is scheduled, all stored data will be removed and the
device will eventually be unprovisioned. Until then, the device is not
completely unprovisioned, thus calling bt_mesh_prov_enable,
bt_mesh_provision (and bt_mesh_cdb_create) should not be allowed.

Struct bt_mesh_prov has a reset callback stating that after this
callback is called, the device has been reset and can be re-provisioned
again. Also, bt_mesh_reset API description states that after calling
bt_mesh_reset API, the device needs to reenable the provisioning layer
to be provisioned again. But this is not correct since the settings has
to be cleared before the device can be reprovisionined.

This commit makes bt_mesh_reset flush pending settings so that the
device can be reprovisioned immediately and the API will behave as
written in the description.

The same applies to bt_mesh_cdb_clear.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
Bluetooth: Mesh: Call bt_mesh_reset in cfg srv from workqueue

Struct bt_mesh_send_cb.end is called from the mesh advertiser thread.
bt_mesh_reset now requires more stack size, thus can't be called from
the advertiser thread. k_work instance requires less memory than needed
for bt_mesh_reset.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
